packages/castle/src/castle.ts

```diff
@@ -42,10 +43,21 @@ export const createCastle = (config: CastleConfig): Castle => {
   const schemaRegistry = new SchemaRegistry(config.schemaRegistry);
   const kafka = new AvroKafka(schemaRegistry, new Kafka(config.kafka), config.topicsAlias);
   const producer = kafka.producer(config.producer);
-  const consumers: CastleConsumer[] = config.consumers.map(config => ({
-    instance: kafka.consumer(config),
-    config,
-  }));
+  const consumers: CastleConsumer[] = config.consumers.map(config => {
+    let consumerConfig: CastleConsumerConfig;
+    if (config.eachSizedBatch) {
+      if (config.eachBatch) {
+        throw new Error('Invalid configuration, please choose one of eachSizedBatch and eachBatch');
+      }
+      consumerConfig = eachSizedBatch(config);
+    } else {
+      consumerConfig = config;
+    }
+    return {
+      instance: kafka.consumer(consumerConfig),
+      config: consumerConfig,
+    };
```

Comment:
Whenever I see mutation myself I think can be an opportunity for a new function. Also having a lot of optional params on that object is not good I think.
Instead of the possible three EachMessage, EachBatch, EachSizedBatch, the types express a lot of possible permutations between all of those. Granted this is a problem with Kafkajs itself, but we can at least not add to that issue.

What do you think if we split our types into three:

```typescript
interface CastleTopicSubscribeBase {
  // ...
}
interface CastleTopicSubscribeEachMessage extends CastleTopicSubscribeBase {
  // Notice that its not optional
  eachMessage: (ctx: CastleEachMessagePayload<T>) => Promise<void>;
}

interface CastleTopicSubscribeEachBatch extends CastleTopicSubscribeBase {
  eachBatch: (ctx: CastleEachBatchPayload<T>) => Promise<void>;
}

interface CastleTopicSubscribeEachSizedBatch extends CastleTopicSubscribeBase {
  eachSizedBatch: (ctx: CastleEachBatchPayload<T>) => Promise<void>;
  batchSize: number;
}

export type CastleTopicSubscribe = CastleTopicSubscribeEachMessage | CastleTopicSubscribeEachBatch | CastleTopicSubscribeEachSizedBatch;
```

We shouldn't need the runtime checks then and throw exceptions as we could never assign the wrong combination.
You would get a static check for the `batchSize` argument as well.

Or if we could we could write a couple of type guards and throw an exception if we mess up on any of the other permutations - like `eachMessage` + `eachSizedBatch`, or `eachMessage` and batchSize for example.

And we shouldn't need `DEFAULT_MAX_BATCH_SIZE` as it wouldn't make sense - you wouldn't use eachSizedBatch if you specifically wanted to alter the size of the batch right? You'd use `eachBatch`.

What do you think? You mentioned that its not _really_ blocking you so I wanted to spend a bit of time to get this right.

Reply:
I agree to everything.

It's not blocking indeed, I have a closed source version of this PR running in prod already :-)

Reply:
I tried the types as you provided and some other variations.
Using interfaces, TS complains that an interface cannot extend an intersection, for `CastleConsumerConfig`
Using types, with or without the `never` trick, TS is confused enough that it expects some of the batch consumers in the tests to provide `eachMessage`, and it looses its guessing ability for `T` which makes all consumers needing to provide their type in the callback function to quiet the "implicity has any type" errors. 

Interesting point re typeguard, not used them so far, I'll have a look.

I could make `eachSizedBatch` an object `{ maxSize, handler}` but I feel that makes configuration compositiion a little bit harder and would break the implicit style of the configuration more than `batchSize` would.

For `DEFAULT_MAX_BATCH_SIZE`: I could make it that if a batchSize is no provided, `eachSizedBatch` is moved to `eachBatch` directly.

Reply:
let me have a go at it

Reply:
#10 what do you think @nicolaslt ?